### PR TITLE
mark Instrumentation Key APIs Obsolete

### PR DESCRIPTION
**Pull request by TimothyMothra**
#2560 

## Changes
- mark PublicAPIs using Instrumentation Key as Obsolete.

image

### CODE REVIEW

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/TelemetryConfiguration.cs [hunk 1]
```diff
@@ -74,14 +74,17 @@ static TelemetryConfiguration()
...
         public TelemetryConfiguration() : this(string.Empty, null)
+#pragma warning restore CS0618 // Type or member is obsolete
         {
         }
 
         /// <summary>
         /// Initializes a new instance of the TelemetryConfiguration class.
         /// </summary>
         /// <param name="instrumentationKey">The instrumentation key this configuration instance will provide.</param>
+        [Obsolete("InstrumentationKey based global ingestion is being deprecated. Transition to using connection strings for data ingestion. https://aka.ms/MigrateToConnectionString")]
```
Comment (TimothyMothra) [inline BASE/src/Microsoft.ApplicationInsights/Extensibility/TelemetryConfiguration.cs:87]: @cartersocha, need to review the messaging here

**Review by reyang — APPROVED** [review]
LGTM.

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/TelemetryConfiguration.cs (hunk 1 above)
Comment (TimothyMothra) [reply BASE/src/Microsoft.ApplicationInsights/Extensibility/TelemetryConfiguration.cs:87]: Spoke to Carter offline, he recommended using this link:
https://azure.microsoft.com/en-us/updates/technical-support-for-instrumentation-key-based-global-ingestion-in-application-insights-will-end-on-31-march-2025/

File: NETCORE/src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs [hunk 2]
```diff
@@ -44,6 +44,7 @@ public static IApplicationBuilder UseApplicationInsightsExceptionTelemetry(this
         /// <param name="services">The <see cref="IServiceCollection"/> instance.</param>
         /// <param name="instrumentationKey">Instrumentation key to use for telemetry.</param>
         /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        [Obsolete("InstrumentationKey based global ingestion is being deprecated. Transition to using connection strings for data ingestion. https://aka.ms/MigrateToConnectionString")]
```
Comment (cijothomas) [inline NETCORE/src/Microsoft.ApplicationInsights.AspNetCore/Extensions/ApplicationInsightsExtensions.cs:47]: for an user of this API, this message is not telling how to migrate. The link is for an overall migration approach.

**Comment by cijothomas** [comment]
@TimothyMothra I think we should offer specific alternate for each method we are Obsoleting.
For eg: service.AddAI(ikey) -- obsolete this, and suggest user to use service.AddAI(new AiOptions(){connectionstring}) ? Or create a github issue, document the migration there, and have obsolete message link to it. The github issue can further link to the aka.ms link to overall migration.

**Comment by TimothyMothra** [comment]
@cijothomas I did both.
Updated the Obsolete message to point to our suggested alternative.
Also updated the linked issue with a full summary of these changes and as well as the relevant links to public docs.
Please re-review when you have a moment.

File: NETCORE/src/Microsoft.ApplicationInsights.WorkerService/ApplicationInsightsExtensions.cs [hunk 3]
```diff
@@ -21,6 +21,7 @@ public static partial class ApplicationInsightsExtensions
         /// <param name="services">The <see cref="IServiceCollection"/> instance.</param>
         /// <param name="instrumentationKey">Instrumentation key to use for telemetry.</param>
         /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        [Obsolete("InstrumentationKey based global ingestion is being deprecated. Use an AddApplicationInsightsTelemetryWorkerService() overload to set ApplicationInsightsServiceOptions.ConnectionString. See https://github.com/microsoft/ApplicationInsights-dotnet/issues/2560 for more details.")]
```
Comment (cijothomas) [inline NETCORE/src/Microsoft.ApplicationInsights.WorkerService/ApplicationInsightsExtensions.cs:24]: the message is not very clear to me.
There is simply no other overload which takes ConnectionString. So suggesting to use an overload without being very clear which overload we are talking about can be confusing.

Comment (TimothyMothra) [reply NETCORE/src/Microsoft.ApplicationInsights.WorkerService/ApplicationInsightsExtensions.cs]: I reviewed all the messages and added additional details.

I reworded this specific message to say `Use an AddApplicationInsightsTelemetryWorkerService() overload to set ApplicationInsightsServiceOptions.ConnectionString via Action<ApplicationInsightsServiceOptions>`

**Review by cijothomas — APPROVED** [review]

**Pull request merged**